### PR TITLE
[core][state] Add "humanify" feature to StateSchema 

### PR DESCRIPTION
**Pull request by ProjectsByJackHe**
## Why are these changes needed?

This PR introduces a framework to address this issue: https://github.com/ray-project/ray/issues/31876.

Essentially, we add a humanify() method to the base stateSchema class, and any subclasses would provide relevant format_fn as a metadata argument to any of its fields, and the humanify() method would aggregate the output from the lambdas. 

This PR is meant to introduce the general framework, and any additions (new format_fn) can be added by request. 

## Related issue number

#31876 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/state/common.py [hunk 1]
```diff
@@ -561,7 +594,11 @@ class TaskState(StateSchema):
     #: The time when the task is created. A Unix timestamp in ms.
     creation_time_ms: Optional[int] = state_column(detail=True, filterable=False)
     #: The time when the task starts to run. A Unix timestamp in ms.
-    start_time_ms: Optional[int] = state_column(detail=True, filterable=False)
+    start_time_ms: Optional[int] = state_column(
+        detail=True,
+        filterable=False,
+        metadata={"format_fn": lambda x: datetime.datetime.fromtimestamp(x / 1000)},
```
Comment (rickyyx) [inline python/ray/experimental/state/common.py:600]: nit: why don't we just `format_fn` a standalone key-value arg, instead of embedding it in the metadata struct?

File: python/ray/experimental/state/common.py [hunk 2]
```diff
@@ -506,6 +530,15 @@ class ClusterEventState(StateSchema):
 class TaskState(StateSchema):
     """Task State"""
 
+    @classmethod
+    def humanify(cls, state: dict) -> dict:
+        if not cls.is_valid_state(state):
```
Comment (rickyyx) [inline python/ray/experimental/state/common.py:535]: Wondering if we could move this logic to the parent state, the Abstract state class.

**Review by rickyyx — COMMENTED** [review]
Approach looks good to me in general! Have some small questions. Super excited to get this merged. 

@rkooo567 could you also PTAL if this is something you expect for it to work?

Also, after we are aligned with the approach, let's add humanify fn to all the fields, we could have a few default functions that humanify common units (e.g., ms to human readable time, bytes to human readable size etc)

File: python/ray/experimental/state/state_cli.py [hunk 3]
```diff
@@ -287,14 +287,9 @@ def format_get_api_output(
     if not state_data or len(state_data) == 0:
         return f"Resource with id={id} not found in the cluster."
 
-    if schema == TaskState and format == AvailableFormat.YAML:
```
[the remark this answers is not in the document]
Comment (ProjectsByJackHe) [reply python/ray/experimental/state/state_cli.py:290]: right, exactly! PR for that was merged already.

File: python/ray/experimental/state/common.py (hunk 2 above)
Comment (ProjectsByJackHe) [reply python/ray/experimental/state/common.py:535]: We could! I originally was thinking each subclass would have to define its own humanify logic. But after the design has been flushed out a bit more, there isn't a need to re-define in the subclasses. I can push an update!

File: python/ray/experimental/state/common.py (hunk 1 above)
Comment (rkooo567) [reply python/ray/experimental/state/common.py:596]: Looks like detail and filterable fields just use metadata under the hood though? Maybe we can use the same pattern?
```

def state_column(*, filterable: bool, detail: bool = False, **kwargs):
    """A wrapper around dataclass.field to add additional metadata.

    The metadata is used to define detail / filterable option of
    each column.

    Args:
        detail: If True, the column is used when detail == True
        filterable: If True, the column can be used for filtering.
        kwargs: The same kwargs for the `dataclasses.field` function.
    """
    m = {"detail": detail, "filterable": filterable}
    if "metadata" in kwargs:
        kwargs["metadata"].update(m)
    else:
        kwargs["metadata"] = m
    return field(**kwargs)
```

**Review by rkooo567 — COMMENTED** [review]
This is really nice! We should use it for other fields like memory notation.

File: python/ray/experimental/state/common.py [hunk 4]
```diff
@@ -197,8 +198,31 @@ class State(StateSchema):
...
+    Subclasses should override this method, providing logic to convert its own fields
+    to something human readable, packaged and returned in a dict.
+
+    Each field that wants to be humanified should include a 'format_fn' key in its
+    metadata dictionary.
     """
 
+    @classmethod
+    def humanify(cls, state: dict) -> dict:
+        """Return a human readable state object ready for printing. Subclasses should override this method."""
```
[the remark this answers is not in the document]
Comment (rickyyx) [reply python/ray/experimental/state/common.py]: Nice - i think that will be easier for extending new states (if we ever do so)

**Comment by rickyyx** [comment]
Let's add a few default humanify functions for some common measurement, i.e. time/memory at least. 

Super excited for this change, this is gonna improve the UX a lot. 

Let me know if you need help resolving the conflcit.

**Comment by ProjectsByJackHe** [comment]
> Let's add a few default humanify functions for some common measurement, i.e. time/memory at least.
> 
> Super excited for this change, this is gonna improve the UX a lot.
> 
> Let me know if you need help resolving the conflcit.

For sure! Apologies for the hiatus, I was cramming for exams (graduating soon lmao).

I can look into adding some more default converter functions for some of the fields in TaskState. Let me know if that's enough scope for this PR or if we need to include Actors... etc.

File: python/ray/experimental/state/common.py (hunk 1 above)
Comment (ProjectsByJackHe) [reply python/ray/experimental/state/common.py:596]: Ahh I see. When you say use the same pattern, do you mean if we explicitly include a metadata parameter, we don't specify detail=?, filterable=? as separate arguments, but rather, as additional keys? Or refactor state_column to not accept detail / filterable in general and have some default keys in metadata underneath the hood if those keys are not specified?

**Comment by rickyyx** [comment]
> > Let's add a few default humanify functions for some common measurement, i.e. time/memory at least.
> > Super excited for this change, this is gonna improve the UX a lot.
> > Let me know if you need help resolving the conflcit.
> 
> For sure! Apologies for the hiatus, I was cramming for exams (graduating soon lmao).
> 
> I can look into adding some more default converter functions for some of the fields in TaskState. Let me know if that's enough scope for this PR or if we need to include Actors... etc.

Sure - we could start with TaskState in the PR (or feel free to expand to other states).

I think we probably don't need that many defaults, i can think of: 
- duration converter
- timestamp
- bytes converter

^ probably just 3 of those?

**Comment by ProjectsByJackHe** [comment]
> > > Let's add a few default humanify functions for some common measurement, i.e. time/memory at least.
> > > Super excited for this change, this is gonna improve the UX a lot.
> > > Let me know if you need help resolving the conflcit.
> > 
> > 
> > For sure! Apologies for the hiatus, I was cramming for exams (graduating soon lmao).
> > I can look into adding some more default converter functions for some of the fields in TaskState. Let me know if that's enough scope for this PR or if we need to include Actors... etc.
> 
> Sure - we could start with TaskState in the PR (or feel free to expand to other states).
> 
> I think we probably don't need that many defaults, i can think of:
> 
> * duration converter
> * timestamp
> * bytes converter
> 
> ^ probably just 3 of those?

Sounds good! I pulled upstream/master on my local branch to resolve merge conflicts. I resolved them, and pushed. However, I am noticing that the latest version of Ray is quite buggy (as expected?). For instance, when I tried to start a cluster via 'ray start --head' I get a circular import error. I can work off a previous version of the codebase to implement and test my changes, and push those changes into this PR, let me know if I am doing something wrong!

**Comment by rickyyx** [comment]
Have you tried running `pip install -e python/ -v` after checking out to a newer commit?

**Comment by ProjectsByJackHe** [comment]
> 

No, I didn't run that command when checked out the new commit. I just ran it, had to upgrade my installed bazel version to > 5.4.0, but then now I am getting this error:

Screen Shot 2023-04-30 at 5 49 49 PM

**Comment by rickyyx** [comment]
Yeah, I would try bazel == 5.4.0 as a quick test.

**Comment by ProjectsByJackHe** [comment]
> Yeah, I would try bazel == 5.4.0 as a quick test.

Awesome! That worked :)
Had some trouble installing bazel on Apple M1 haha

Updated PR such that all 60 unit tests in state_cli passes, and manual testing holds up.

File: python/ray/experimental/state/common.py [hunk 5]
```diff
@@ -516,6 +548,38 @@ class ClusterEventState(StateSchema):
     custom_fields: Optional[dict] = state_column(filterable=False, detail=True)
 
 
+class Humanify:
+    """A class containing default methods to convert units into a human readable string."""
+
+    def timestamp(x: float):
+        """Converts miliseconds to a datetime object."""
+        return datetime.datetime.fromtimestamp(x / 1000)
```
[the remark this answers is not in the document]
Comment (ProjectsByJackHe) [reply python/ray/experimental/state/common.py:556]: Screen Shot 2023-05-02 at 10 04 12 PM

File: python/ray/experimental/state/common.py [hunk 6]
```diff
@@ -516,6 +548,38 @@ class ClusterEventState(StateSchema):
...
+        for event in events:
+            if "created_ms" in event:
+                event["created_ms"] = Humanify.timestamp(event["created_ms"])
+            if "start_time" in event:
+                event["start_time"] = Humanify.timestamp(event["start_time"])
+            if "end_time" in event:
+                event["end_time"] = Humanify.timestamp(event["end_time"])
+        return events
+
+    def profiliing_data(data: dict):
```
[the remark this answers is not in the document]
Comment (ProjectsByJackHe) [reply python/ray/experimental/state/common.py:577]: I see, thanks for letting me know. Will get rid of it in the next update.

File: python/ray/experimental/state/state_cli.py [hunk 7]
```diff
@@ -292,10 +292,17 @@ def format_get_api_output(
...
     if not isinstance(state_data, list):
         state_data = [state_data]
-    state_data = [dataclasses.asdict(state) for state in state_data]
-    return output_with_format(state_data, schema=schema, format=format, detail=True)
+
+    human_readable_state_data = [
+        dataclasses.asdict(schema.humanify(state)) for state in state_data
+    ]
+
+    return output_with_format(
```
[the remark this answers is not in the document]
Comment (ProjectsByJackHe) [reply python/ray/experimental/state/state_cli.py:303]: Sure, I wasn't certain if we wanted to humanify ALL state_data that calls `output_with_format`, but if that's the case, then it would be cleaner if we did the humanify logic in output_with_format.

File: python/ray/experimental/state/common.py [hunk 8]
```diff
@@ -516,6 +548,38 @@ class ClusterEventState(StateSchema):
...
+class Humanify:
+    """A class containing default methods to convert units into a human readable string."""
+
+    def timestamp(x: float):
+        """Converts miliseconds to a datetime object."""
+        return datetime.datetime.fromtimestamp(x / 1000)
+
+    def memory(x: int):
+        """Converts raw bytes to megabytes"""
+        return str(x / 1e6) + " MB"
```
[the remark this answers is not in the document]
Comment (ProjectsByJackHe) [reply python/ray/experimental/state/common.py:560]: For sure! Added additional helpers for extra configurability.

File: python/ray/experimental/state/common.py [hunk 9]
```diff
@@ -516,6 +548,38 @@ class ClusterEventState(StateSchema):
...
+
+    def duration(x: int):
+        """Converts miliseconds to a human readable duration."""
+        return str(datetime.timedelta(milliseconds=x))
+
+    def events(events: List[dict]):
+        """Converts a list of events into a human readable format."""
+        for event in events:
+            if "created_ms" in event:
```
[the remark this answers is not in the document]
Comment (ProjectsByJackHe) [reply python/ray/experimental/state/common.py:569]: Correct me if I am wrong, I believe L1390-L1407 in common.py takes as input a TaskEvent protobuf object, but the Humanify.events helper takes as input a list of dictionaries (output of L1388), which are missing the "{...}_ts" keys.

File: python/ray/experimental/state/common.py [hunk 10]
```diff
@@ -516,6 +548,38 @@ class ClusterEventState(StateSchema):
     custom_fields: Optional[dict] = state_column(filterable=False, detail=True)
 
 
+class Humanify:
+    """A class containing default methods to convert units into a human readable string."""
+
+    def timestamp(x: float):
```
[the remark this answers is not in the document]
Comment (ProjectsByJackHe) [reply python/ray/experimental/state/common.py:553]: Sure! Added a few helpers in test_sanity_check to parse output_string for a specific key and verify the value is of correct format. As a future TODO, it would be nice to figure out why json.loads(output_from_runner) does not work, and I wouldn't have to use a hacky regular expression solution.

**Comment by rickyyx** [comment]
let me know when it's ready for another review!

**Comment by ProjectsByJackHe** [comment]
> let me know when it's ready for another review!

For sure! So I've made a few changes and responded to your comments. 

My changes:
- Add unit test to check correct format of the timestamp after humanifying. 
- Moved humanify to output_with_format
- Added 2 more helpers in Humanify class to convert to kilobytes / gigabytes, the developer can choose which helper they want to use in format_fn.

Regarding the comment about finding a more generic way to parse for timestamp in events, correct me if I am wrong, I believe L1390-L1407 in common.py takes as input a TaskEvent protobuf object, but the Humanify.events helper takes as input a list of dictionaries (output of L1388), which are missing the "{...}_ts" keys.

Also, what are your thoughts on the way I've been doing the unit tests? Seems like there is no clean way to parse a state_cli output from runner as 'json.loads()' doesn't quite work and so I've had to implement get_key using regex.

File: python/ray/experimental/state/common.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (rickyyx) [reply python/ray/experimental/state/common.py:576]: I see, so what are the input events here? Seems we only need to support TaskState.events right? Does it still have the `crated_ms` field? 

Or we could the `events` format function altogether if the format is too customized. I don't have a strong opinion about this,

**Review by rickyyx — CHANGES_REQUESTED** [review]
Good job on the e2e testing, left a few comments. We are pretty close to merging this I beleive!

<div><br/>via <a href="https://giphy.com/gifs/yJFeycRK2DB4c">GIPHY</a></div>

File: python/ray/experimental/state/common.py [hunk 11]
```diff
@@ -516,6 +547,37 @@ class ClusterEventState(StateSchema):
...
+        """Converts miliseconds to a datetime object."""
+        return datetime.datetime.fromtimestamp(x / 1000)
+
+    def to_MiB(x: int):
```
[the remark this answers is not in the document]
Comment (ProjectsByJackHe) [reply python/ray/experimental/state/common.py:557]: Yeah, that's a much smarter solution 👍

File: python/ray/tests/test_state_api.py [hunk 12]
```diff
@@ -1974,6 +1976,53 @@ def verify_output(cmd, args: List[str], necessary_substrings: List[str]):
         )
     )
 
+    # json.loads(output_str_from_runner) does not work. So we use regular expression to parse the output and get a key.
```
[the remark this answers is not in the document]
Comment (ProjectsByJackHe) [reply python/ray/tests/test_state_api.py:1979]: Ahh I see, yes that's much easier. Will add them.

File: python/ray/experimental/state/state_cli.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ProjectsByJackHe) [reply python/ray/experimental/state/state_cli.py:304]: Hmm, I thought I reverted this change already? Perhaps I forgot to push. Let me try again.

File: python/ray/experimental/state/common.py [hunk 13]
```diff
@@ -193,8 +196,36 @@ class State(StateSchema):
...
+    to something human readable, packaged and returned in a dict.
+
+    Each field that wants to be humanified should include a 'format_fn' key in its
+    metadata dictionary.
     """
 
+    @classmethod
+    def humanify(cls, state: dict) -> dict:
+        """Convert the given state object into something human readable."""
+        if not cls.is_valid_state(state):
```
[the remark this answers is not in the document]
Comment (ProjectsByJackHe) [reply python/ray/experimental/state/common.py:213]: I assumed humanify was an atomic operation. Since python doesn't enforce types, I was thinking that we only mutate the data (state_dict) if it was a valid data class (containing all necessary fields) passed in.

What do you think?

File: python/ray/experimental/state/common.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ProjectsByJackHe) [reply python/ray/experimental/state/common.py:576]: the input events here right now is only used for TaskState.events, it still has the created_ms field. 

My thinking was Humanify would be modified according to use case (with feedback from users), and future contributors will have a better sense of how they want to interpret the CLI output. I just made a wild guess that TaskState.events.created_ms might benefit from being humanified, and left this implementation as a future reference in case maybe another field gets added to TaskState.events and contributors can adjust the code to format it however they need.

File: python/ray/experimental/state/common.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ProjectsByJackHe) [reply python/ray/experimental/state/common.py:553]: UPDATE: I removed the complex regex in favor of simpler testing methods from feedback.

**Comment by ProjectsByJackHe** [comment]
When I followed the DCO instructions to do a git rebase HEAD~18 to fix signing off commits issue, something like this happened. 

Sorry!

This is where we left off: https://github.com/ray-project/ray/pull/35059

**Comment by rickyyx** [comment]
we should probably don't do rebase, merge is better.
